Adding solutions

FKN237 opened the pull request:
---
name: solution review 
about: A template PR for code review with a checklist
---

<!--
  make this PR easy to find:

  - assign: yourself
  - link it to the corresponding challenge issue on the project board
-->

<!-- describe the challenge you will solve -->

## Behavior

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

## Strategy

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

FKN237 committed:
Add files via upload

FKN237 committed:
Add files via upload

FKN237 committed:
Update fizz_buzz.py

FKN237 committed:
Update fizz_buzz.py

matopcheg reviewed (changes requested):
Amazing, Nelson! You did a very good job. I did look on some checkboxes and noticed that some of them are not satisfied. If you can make some changes, it would be nice.
- You need to have 3 or more doctests in function file
- Make a module dosctring for test and function file (just write a summary docstring in the beggining of your files)
- Add defensive assertions. 
- Correct type of annotations in function. For example, in count_vowels you can add def count_vowels(s: str) -> int:

If you need any help - feel free to ask me or Nagham and Salem!

FKN237 commented:
> Amazing, Nelson! You did a very good job. I did look on some checkboxes and noticed that some of them are not satisfied. If you can make some changes, it would be nice.
> 
> * You need to have 3 or more doctests in function file
> * Make a module dosctring for test and function file (just write a summary docstring in the beggining of your files)
> * Add defensive assertions.
> * Correct type of annotations in function. For example, in count_vowels you can add def count_vowels(s: str) -> int:
> 
> If you need any help - feel free to ask me or Nagham and Salem!

Alright 👌 thank you Matvii. I tried adding the module doctring, but when I do that, I keep having python formatting issues

FKN237 committed:
Update fizz_buzz.py

FKN237 committed:
Update fizz_buzz.py

FKN237 committed:
Update test_fizz_buzz.py

FKN237 committed:
Update test_fizz_buzz.py

adding a module doccstring with some path modifications to avoid formatting issues

FKN237 committed:
Update count_vowels.py

added module docstring, and defensive assertions

FKN237 committed:
Update test_count_vowels.py

Added a module docstring and defensive assertions

FKN237 committed:
Update count_vowels.py

final touches

FKN237 committed:
Update test_count_vowels.py

FKN237 committed:
Update reverse_string.py

FKN237 committed:
Update count_vowels.py

FKN237 committed:
Update test_reverse_string.py

final touches

FKN237 committed:
Update fizz_buzz.py

final touches

FKN237 committed:
Update fizz_buzz.py

FKN237 committed:
Update fizz_buzz.py

FKN237 committed:
Update test_fizz_buzz.py

matopcheg committed:
Merge branch 'adding_solutions' of https://github.com/MIT-Emerging-Talent/ET6-foundations-group-20 into adding_solutions

matopcheg committed:
ruff formatting

matopcheg reviewed (approved):
Good job!

matopcheg committed:
Merge branch 'main' into adding_solutions

matopcheg committed:
Merge branch 'main' into adding_solutions
